### PR TITLE
Fix Nuget Cuda pipeline package pipeline

### PR DESCRIPTION
**Pull request by jchen351**
### Description

This PR adding protoc.exe to make the Nuget Cuda Pipleine, which also allowing it to get build Java for various CUDA version

### CODE REVIEW

**Comment by tianleiwu** [comment]
It is good that we support cuda 12 as major (instead of cuda 11) in nightly going forward.

Any reason that stable diffusion test failed?

File: tools/ci_build/github/windows/extract_nuget_files_gpu.ps1 [hunk 1]
```diff
@@ -4,21 +4,21 @@
 # This file is used by Zip-Nuget-Java Packaging Pipeline
 New-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo\nuget-artifacts -ItemType directory
```
Comment (snnn) [inline tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:5]: Space only changes?

Comment (jchen351) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:5]: line 32 is the only change. Some others are just white space changes.

**Review by snnn — DISMISSED** [review]

**Comment by jchen351** [comment]
> It is good that we support cuda 12 as major (instead of cuda 11) in nightly going forward.
> 
> Any reason that stable diffusion test failed?

It is not failed anymore?

File: tools/ci_build/github/windows/extract_nuget_files_gpu.ps1 [hunk 2]
```diff
@@ -29,14 +29,14 @@ Foreach-Object {
 New-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo -ItemType directory
 
 Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo\nuget-artifacts\onnxruntime-win-x64-cuda-*\lib\* -Destination $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo
-Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\extra-artifact\protoc.exe $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo
+dotnet tool install --global protoc
```
Comment (snnn) [inline tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: Pin a version here.

Comment (snnn) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: We should use the same version as what we have in cmake/deps.txt

Comment (jchen351) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: 3.6.1?

File: tools/ci_build/github/azure-pipelines/templates/win-ci.yml [hunk 3]
```diff
@@ -384,16 +387,3 @@ stages:
...
             arguments: '--config RelWithDebInfo --use_binskim_compliant_compile_flags --enable_lto --disable_rtti --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --test --enable_onnx_tests  $(TelemetryOption) '
             workingDirectory: '$(Build.BinariesDirectory)'
-
-        - ${{ if eq(parameters.buildJava, 'true') }}:
-          - template: make_java_win_binaries.yml
-            parameters:
-              msbuildPlatform: ${{ parameters.msbuildPlatform }}
-              java_artifact_id: ${{ parameters.java_artifact_id }}
-
-          - task: PublishBuildArtifacts@1
```
Comment (jchen351) [inline tools/ci_build/github/azure-pipelines/templates/win-ci.yml:394]: Publish this before testing

File: tools/ci_build/github/azure-pipelines/stages/java-cuda-packaging-stage.yml [hunk 4]
```diff
@@ -7,11 +7,11 @@ parameters:
...
 
 stages:
-- stage: Java_GPU_Packaging
+- stage: Jar_Packaging_GPU
   dependsOn:
-  - Linux_C_API_Packaging_Combined_CUDA
-  - Windows_Packaging_CUDA
-  - Windows_Packaging_TensorRT
+  - Linux_C_API_Packaging_GPU
+  - Windows_Packaging_CUDA_Testing
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/stages/java-cuda-packaging-stage.yml:13]: Why is this change needed?

Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/stages/java-cuda-packaging-stage.yml:13]: Because build and publishing Java jar steps were moved to Windows_Packaging_CUDA_Testing stage where GPU machine is used. The same issue regarding gradlew I just told you ealier. `Windows_Packaging_CUDA` is run on CPU machine.

File: tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml [hunk 5]
```diff
@@ -8,13 +8,11 @@ parameters:
...
 - name: buildJava
   type: boolean
-  default: false
 - name: buildNodejs
   type: boolean
-  default: false
 
 stages:
-- stage: Linux_C_API_Packaging_Combined_CUDA
+- stage: Linux_C_API_Packaging_GPU
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:15]: So, right now the stage has two jobs. Your this change split them to two stages. How does it help? Why is this change needed?

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:15]: Why do you need to change this name from "Linux_C_API_Packaging_Combined_CUDA" to "Linux_C_API_Packaging_GPU" ?

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:15]: Oh, now I understand. 
Before your change, this file has one stage which contains 3 jobs.
After your change, this file has two stages. The first one contains two jobs. The second one contains one job. The second stage depends on the first stage. 
Why is it better?

File: tools/ci_build/github/windows/extract_nuget_files_gpu.ps1 [hunk 6]
```diff
@@ -29,14 +29,14 @@ Foreach-Object {
 New-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo -ItemType directory
 
 Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo\nuget-artifacts\onnxruntime-win-x64-cuda-*\lib\* -Destination $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo
-Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\extra-artifact\protoc.exe $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo
+dotnet tool install --global protoc --version 3.6.1
```
Comment (snnn) [inline tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: Should not be 3.6.1. Should be the one we use in cmake/deps.txt.

Comment (snnn) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: The package you choose doesn't seem to be published by Google.  And it is out of maintain.

Comment (snnn) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:32]: I think you should use google.protobuf.tools. I checked, the DLLs are signed by Google.

**Comment by snnn** [comment]
The PR's title says: "Enable Nuget Cuda pipeline package publishing nightly", but how is enabled through the changes? Did you add a trigger to a pipeline, or did you fix some issues to make it possible?

File: tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml (hunk 5 above)
Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:15]: The reason doing that because downstream stage like Jar_Packaging_GPU, NuGet_Packaging_GPU and Nodejs_Packaging doesn't depend on the 3rd job, which is now the second stage, this will speed up the whole pipeline allow downstream to run without waiting on the the second stage.

File: tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml [hunk 7]
```diff
@@ -147,10 +145,11 @@ stages:
         condition: 'succeeded'
     - template: ../templates/clean-agent-build-directory-step.yml
   # Linux CUDA Combined Testing and Publishing
+- stage: Linux_Packaging_combined_CUDA
```
[the remark this answers is not in the document]
Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:148]: because I combine the previous 2 jobs into single stage.

File: tools/ci_build/github/azure-pipelines/templates/win-ci.yml [hunk 8]
```diff
@@ -307,9 +307,12 @@ stages:
       - template: component-governance-component-detection-steps.yml
         parameters :
           condition : 'succeeded'
-  - ${{ if contains(parameters.ort_build_pool_name, 'GPU') }}:
+- ${{ if contains(parameters.ort_build_pool_name, 'GPU') }}:
```
[the remark this answers is not in the document]
Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/templates/win-ci.yml:310]: actually, this change isn't necessary anymore. I was planning to make downstream stages to run without waiting for the GPU testing.

File: tools/ci_build/github/azure-pipelines/stages/java-cuda-packaging-stage.yml (hunk 4 above)
Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/stages/java-cuda-packaging-stage.yml]: I rolled back previous changes, this is not needed anymore

File: tools/ci_build/github/azure-pipelines/templates/c-api-cpu.yml [hunk 9]
```diff
@@ -511,7 +511,7 @@ stages:
   - Windows_CI_GPU_DML_Dev
   - Windows_CI_GPU_DML_Dev_arm64
   - Linux_C_API_Packaging_CPU
-  - Linux_C_API_Packaging_Combined_CUDA
+  - Linux_C_API_Packaging_GPU
```
[the remark this answers is not in the document]
Comment (jchen351) [reply tools/ci_build/github/azure-pipelines/templates/c-api-cpu.yml:514]: The reason doing that because downstream stage like Jar_Packaging_GPU, NuGet_Packaging_GPU and Nodejs_Packaging doesn't depend on the 3rd job, which is now the second stage, this will speed up the whole pipeline allow downstream to run without waiting on the the second stage.

File: tools/ci_build/github/windows/extract_nuget_files_gpu.ps1 [hunk 10]
```diff
@@ -29,7 +29,9 @@ Foreach-Object {
 New-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo -ItemType directory
 
 Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo\nuget-artifacts\onnxruntime-win-x64-cuda-*\lib\* -Destination $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\RelWithDebInfo
-Copy-Item -Path $Env:BUILD_BINARIESDIRECTORY\extra-artifact\protoc.exe $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo
+dotnet new console
+dotnet add package Google.Protobuf.Tools --version 3.21.12
+Copy-Item -Path $Env:USERPROFILE\.nuget\packages\google.protobuf.tools\**\tools\windows_x64\protoc.exe $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build\RelWithDebInfo
```
Comment (snnn) [inline tools/ci_build/github/windows/extract_nuget_files_gpu.ps1:34]: If it is in USERPROFILE folder, the file could come from a different run. Can we put it in the current workspace?

Comment (jchen351) [reply tools/ci_build/github/windows/extract_nuget_files_gpu.ps1]: put it into $Env:BUILD_BINARIESDIRECTORY\RelWithDebInfo\_deps\protobuf-build

File: tools/ci_build/github/azure-pipelines/templates/win-ci.yml [hunk 11]
```diff
@@ -308,92 +308,94 @@ stages:
...
-      timeoutInMinutes: 180
-      steps:
-        - task: mspremier.PostBuildCleanup.PostBuildCleanup-task.PostBuildCleanup@3
-          displayName: 'Clean Agent Directories'
-          condition: always()
-
-        - checkout: self
-          clean: true
-          submodules: none
+    - stage: Windows_Packaging_${{ parameters.stage_name_suffix }}_Testing
```
Comment (jchen351) [inline tools/ci_build/github/azure-pipelines/templates/win-ci.yml:311]: Just making this become a stage

File: tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml [hunk 12]
```diff
@@ -51,14 +49,16 @@ stages:
...
 
-    - template: ../templates/java-api-artifacts-package-and-publish-steps-posix.yml
-      parameters:
-        arch: 'linux-x64'
-        buildConfig: 'Release'
-        artifactName: 'onnxruntime-java-linux-x64-cuda'
-        version: '$(OnnxRuntimeVersion)'
-        libraryName: 'libonnxruntime.so'
-        nativeLibraryName: 'libonnxruntime4j_jni.so'
+    - ${{ if eq(parameters.buildJava, true) }}:
```
Comment (jchen351) [inline tools/ci_build/github/azure-pipelines/stages/nuget-linux-cuda-packaging-stage.yml:52]: Only publish java when BuildJava is true

**Review by snnn — APPROVED** [review]
